### PR TITLE
Allow numerical UIDs in systemctl -M uid@ and similar

### PR DESCRIPTION
**Pull request by keszybz**

### CODE REVIEW

File: src/libsystemd/sd-bus/bus-container.c [hunk 1]
```diff
@@ -50,12 +50,12 @@ int bus_container_connect_socket(sd_bus *b) {
         bus_socket_setup(b);
 
         if (socketpair(AF_UNIX, SOCK_SEQPACKET|SOCK_CLOEXEC, 0, pair) < 0)
-                return -errno;
+                return log_debug_errno(errno, "Failed to create a scoket pair: %m");
```
Comment (poettering) [inline src/libsystemd/sd-bus/bus-container.c]: typo: scoket

File: src/libsystemd/sd-bus/bus-container.c [hunk 2]
```diff
@@ -37,11 +37,11 @@ int bus_container_connect_socket(sd_bus *b) {
 
         r = namespace_open(b->nspid, &pidnsfd, &mntnsfd, NULL, &usernsfd, &rootfd);
         if (r < 0)
-                return r;
+                return log_debug_errno(r, "Failed to open namespace of PID "PID_FMT": %m", b->nspid);
 
         b->input_fd = socket(b->sockaddr.sa.sa_family, SOCK_STREAM|SOCK_CLOEXEC|SOCK_NONBLOCK, 0);
         if (b->input_fd < 0)
-                return -errno;
+                return log_debug_errno(errno, "Failed to create a scoket: %m");
```
Comment (poettering) [inline src/libsystemd/sd-bus/bus-container.c]: typo: scoket

File: src/libsystemd/sd-bus/sd-bus.c [hunk 3]
```diff
@@ -1648,19 +1649,25 @@ static int user_and_machine_equivalent(const char *user_and_machine) {
...
+
+        if (uid == 0 && STR_IN_SET(user_and_machine, ".host", "root@.host", "0@.host"))
+                return true;
+
+        /* Otherwise, we have to figure out our user id and name, and compare things with that. */
+        char buf[DECIMAL_STR_MAX(uid_t)];
+        xsprintf(buf, UID_FMT, uid);
+
+        f = startswith(user_and_machine, buf);
+        if (f && *f == '@')
```
Comment (poettering) [inline src/libsystemd/sd-bus/sd-bus.c]: you don#t check the hostname anymore.

File: src/libsystemd/sd-bus/sd-bus.c [hunk 4]
```diff
@@ -1648,19 +1649,25 @@ static int user_and_machine_equivalent(const char *user_and_machine) {
...
         un = getusername_malloc();
         if (!un)
                 return -ENOMEM;
 
         f = startswith(user_and_machine, un);
-        if (!f)
-                return false;
-
-        return STR_IN_SET(f, "@", "@.host");
+        return f && *f == '@';
```
Comment (poettering) [inline src/libsystemd/sd-bus/sd-bus.c]: ditto

**Comment by keszybz** [comment]
Updated with the typo and the stupid error fixed.

File: src/basic/escape.c [hunk 5]
```diff
@@ -542,3 +543,21 @@ char* shell_maybe_quote(const char *s, ShellEscapeFlags flags) {
...
+char* quote_command_line(char **argv) {
+        _cleanup_free_ char *result = NULL;
+
+        assert(argv);
+
+        char **a;
+        STRV_FOREACH(a, argv) {
+                _cleanup_free_ char *t = NULL;
+
+                t = shell_maybe_quote(*a, SHELL_ESCAPE_EMPTY);
```
Comment (poettering) [inline src/basic/escape.c:556]: oom check missing

File: src/test/test-escape.c [hunk 6]
```diff
@@ -203,6 +203,29 @@ static void test_shell_maybe_quote(void) {
         test_shell_maybe_quote_one("głąb\002\003rząd", SHELL_ESCAPE_POSIX, "$'głąb\\002\\003rząd'");
 }
 
+static void test_quote_command_line_one(char **argv, const char *expected) {
+        _cleanup_free_ char *s;
+
+        s = quote_command_line(argv);
```
Comment (poettering) [inline src/test/test-escape.c]: assert_se() around this?

**Comment by poettering** [comment]
lgtm, except for the oom thing

**Comment by keszybz** [comment]
Updated with the two requested changes.

**Review by bluca — APPROVED** [review]
LGTM and Lennart already approved, so let's merge this

**Pull request merged**